src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,125 @@
+import classNames from 'classnames';
+import React, { useEffect, useRef, useState } from 'react';
+import { Todo } from '../../types/Todo';
+
+export type Props = {
+  todo: Todo;
+  isProcessed: boolean,
+  onUpdate?: (todo: Todo) => Promise<void>;
+  onDelete?: () => Promise<void>;
+};
+
+export const TodoItem: React.FC<Props> = ({
+  todo,
+  isProcessed,
+  onDelete = () => Promise.resolve(),
+  onUpdate = () => Promise.resolve(),
+}) => {
+  const field = useRef<HTMLInputElement>(null);
+  const [editing, setEditing] = useState(false);
+  const [title, setTitle] = useState(todo.title);
+
+  useEffect(() => {
+    if (field.current) {
+      field.current.focus();
+    }
+  }, [editing]);
+
+  function save() {
+    if (isProcessed) {
+      return;
+    }
+
+    const trimmedTitle = title.trim();
+
+    if (trimmedTitle === todo.title) {
+      setEditing(false);
+
+      return;
+    }
+
+    if (trimmedTitle) {
+      onUpdate({ ...todo, title: trimmedTitle })
+        .then(() => setEditing(false))
+        .catch(() => field.current?.focus());
+    } else {
+      onDelete()
+        .then(() => setEditing(false))
+        .catch(() => field.current?.focus());
+    }
+  }
+
+  function onSubmit(event: React.FormEvent) {
+    event.preventDefault();
+    save();
+  }
+
+  return (
+    <div
+      data-cy="Todo"
+      className={classNames('todo', { completed: todo.completed })}
+    >
+      <label className="todo__status-label">
+        <input
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          checked={todo.completed}
+          onChange={() => {
+            onUpdate({ ...todo, completed: !todo.completed })
+              .catch(() => {});
+          }}
```

Comment:
Pls move the function inside `onChange` handler to a separate const and use as a callback here
**PLS DO THE SAME FOR ALL SUCH CASES IN THE PROJECT**